[Jed] Step 3-5 : 리팩토링(구체타입 의존을 추상타입으로 수정)

junu0516 opened the pull request:
## 작업 내용
- [x] 기존의 Rectangle, ColorRectangle, ImageRectangle 등의 구체 타입 사용하던 부분을 각각 별도의 프로토콜 타입을 사용하도록 수정
- [x] ViewController, RectangleViewFactory 내부 로직 프로토콜 타입에 맞춰 수정
- [x] PlaneTest를 수정된 Rectangle 타입에 맞게 수정
    
## 고민과 해결
- Rectangle을 RectangleApplicable이라는 프로토콜로 대신 사용하도록 했습니다.
- ColorRectangle의 경우에는 사각형 속성을 부여하는 RectangleApplicable 프로토콜과, 배경색 속성을 부여하는 RandomColorApplicable 프로토콜을 채택하도록 수정했습니다.
- ImageRectangle의 경우에는 RectangleApplicable과 이미지 속성을 부여하는 CustomImageApplicable 프로토콜을 채택하도록 수정했습니다.
- ViewController과, Plane, RectangleFactory, RectangleViewFactory 내부의 모든 메소드가 기존에는 파라미터 및 리턴타입을 구체적인 Rectangle로 사용하던 것을, 모두 RectangleApplicable로 사용하도록 수정했습니다.
    
- 다음 단계 진행하기 전에, 이전 단계 피드백에서 언급하셨던 __`구체타입이 아닌 프로토콜에 의존하도록 수정`__ 하라고 한 부분을 먼저 수정했습니다. 수정된 양이 좀 많은 것 같아서 부득이하게 먼저 피드백 요청하고자 PR 보냈습니다!

junu0516 committed:
Add: CustomImageApplicable, RandomColorApplicable protocol

junu0516 committed:
Add: RectangleApplicable protocol

junu0516 committed:
Refactor: Rectangle 내부의 구조체 타입들 모두 외부로 이동

junu0516 committed:
Remove: 프로토콜 적용을 위해 기존 Rectangle.swift 삭제

junu0516 committed:
Refactor: RectangleApplicable 프로토콜 속성 정의

- point,alpha에 대해서는 우선 get,set모두 허용

junu0516 committed:
Refactor: CustomStringApplicable, RandomColorApplicable 프로토콜 속성 정의

junu0516 committed:
Refactor: ColorRectangle, ImageRectangle 모두 프로토콜 채택하는 클래스로 수정

junu0516 committed:
Refactor: Plane 내부 Rectangle 타입 사용을 RectangleApplicable 타입 사용으로 모두 수정

junu0516 committed:
Refactor: RectangleFactory 내부 메소드의 리턴타입을 모두 프로토콜타입으로 수정

junu0516 committed:
Refactor: AbstractRectangleView, ColorRectangleView, ImageRectangleView 생성자 파라미터 모두 프로토콜로 받도록 수정

junu0516 committed:
Refactor: ViewController 내부의 Rectangle 타입 사용 모두 RectangleApplicable 타입 사용으로 수정

- 모든 구체타입 추상타입으로 변경

junu0516 committed:
Refactor: RectangleViewFactory 내부의 생성 메소드 불필요한 파라미터 제거

junu0516 committed:
Add: Color.swift 생성, Color 타입 외부 파일로 이동

junu0516 committed:
Revert "Remove: 프로토콜 적용을 위해 기존 Rectangle.swift 삭제"

This reverts commit 6187c0e015680d380ffffefa8d074a4cb2c0a63c.

junu0516 committed:
Refactor: PlaneTest 수정

junu0516 committed:
Chore: 각 파일 목적폴더 별로 분류해서 위치 이동

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -7,7 +7,7 @@ class ViewController: UIViewController{
     private var stylerView: StylerView?
     private var plane: Plane = Plane()
-    private var rectangleDictionary:[Rectangle.Id:UIView] = [:]
+    private var rectangleDictionary:[Id:UIView] = [:]
```
Rectangle 내부에 있는 Id 말고 Rectangle 수준에서 객체를 비교할 수는 없나요?

junu0516 replied:
앗 이전 코멘트에도 키를 모델 수준에서 쓰라 하셔서 해당 부분 계속 고민중입니다!
이후에 이 부분 적용해서 push하겠습니다

godrm commented on DrawingApp/DrawingApp/Model/ColorRectangle.swift:
```diff
@@ -1,40 +1,22 @@
-    }
+class ColorRectangle: RectangleApplicable, RandomColorApplicable, CustomStringConvertible{
     
+    var id: Id
+    var size: Size
+    var point: Point
+    var alpha: Alpha
```
내부 속성은 private var를 우선적으로 고려하시고, 값을 바꿀 수 있더라도 속성 자체를 바꾸기 보다는 메소드로 제공하는 게 좋습니다
그래야 외부에서 속성을 바꾸는 시점도 알 수 있고, 객체 내부에서 값을 안전하게 처리할 수 있습니다. 
그렇지 않으면 struct - let 값 타입을 고려해보세요

junu0516 replied:
넵 접근제어 수정하겠습니다

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -85,7 +89,7 @@ class Plane:CustomStringConvertible{
     }
     
-    subscript(x: Double = 0, y: Double = 0)-> Rectangle?{
+    subscript(x: Double = 0, y: Double = 0)-> RectangleApplicable?{
```
두 가지 의견이 있습니다. 
- isRectangleInsideTheRange() 같은 함수도 rectangle.point.x 내부 속성을 가져오지 않고 `rectangle.isContain(x,y)` 처럼 하위에서 비교할 수 있지 않을까요? 카드 속성을 가져오지 않고 카드 자체를 비교하는 것처럼요
- 일반적인 subscript()는 어떤 용도로 사용하게 될까요? plane에게는 이 subscript(x,y)가 어떤 정보일까요? plane[x,y] 라는게 plane의 특정한 점을 가져오는 것일까요? subscript가 어떤 의도인지 알기 어려운 것 같아서요. 그냥 메소드와 달리 subscript로 제공하는 이유도 궁금합니다.

junu0516 replied:
이 부분은 생각하지 못했네요. 좌표값은 사각형 모델에서 더 직접적으로 사용하다보니 해당 모델 클래스 내부로 이동하는게 더 적절할 것 같습니다.

godrm reviewed (approved):
잘 개선하셨습니다. 하지만 다음 내용들은 꼭 확인해보고 댓글 남겨주세요. 
그걸 보고 개선할 지 머지할 지도 판단해봐야 할 것 같아요

junu0516 commented:
코멘트 감사합니다. 피드백 주신 부분 수정해서 다시 push하겠습니다!(아직 머지는 보류하겠습니다)

junu0516 committed:
Refactor: ColorRectangle, ImageRectangle 속성 private (set) 으로 일괄 변경

junu0516 committed:
Refactor: ViewController의 rectangleDictionary 속성 키를 Rectangle 자체로 사용하도록 수정

- Rectangle 클래스 Hashable 채택
- ColorRectangle, ImageRectangle은 RectangleApplicable이 아닌 Rectangle을 상속받도록 수정

junu0516 committed:
Refactor: Plane 내부 subscript 제거

- 터치된 좌표값 받아서 사각형 모델을 리턴하던 subscript를 메소드로 대체
- 터치된 좌표값이 사각형 범위 안에 있는 지 판별하던 메소드는 RectangleApplicable의 메소드로 이동
- ViewController에서 plane 내부의 사각형 모델 찾는 부분은 subscript가 아닌 메소드 호출로 모두 대체

junu0516 commented:
@godrm  위에 피드백 주셨던 3가지 사항 수정해서 push했습니다. 확인 부탁드립니다!

junu0516 commented:
@godrm  위에 피드백 주셨던 3가지 사항 수정해서 push했습니다. 확인 부탁드립니다!

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -7,7 +7,7 @@ class ViewController: UIViewController{
     private var stylerView: StylerView?
     private var plane: Plane = Plane()
-    private var rectangleDictionary:[Id:UIView] = [:]
+    private var rectangleDictionary:[Rectangle:UIView] = [:]
```
변경된 것 확인했습니다. 
다음 도전꺼리는 여기 Plane, Rectangle, UIView 같은 구체 타입이 아니라 프로토콜에 의존하도록 바꿔보세요. 프로토콜과 의존성 주입해야 하기 때문에 꽤 흥미로운 도전꺼리가 될 겁니다.

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -75,9 +75,11 @@ class ViewController: UIViewController{
+           let rectangle = rectangle as? Rectangle{
+            self.rectangleDictionary[rectangle] = rectangleView
             canvasView.insertSubview(rectangleView, belowSubview: canvasView.generatingButton)
+            print(canvasView.subviews)
```
이 print는 설마 print 구문인가요? 앞으로는 확인하는 것도 디버거를 적극 활용해주세요.

junu0516 replied:
앗 임시로 확인했던 부분인데 제거하는 것을 깜빡했네요! 제거하겠습니다

godrm reviewed (approved):
변경 내용 확인했습니다. 
다음 도전꺼리는 ViewController 나 Plane에서 사용하는 하위 타입들을 구체 타입이 아니라 프로토콜에 의존하도록 바꿔보세요. 프로토콜과 의존성 주입해야 하기 때문에 꽤 흥미로운 도전꺼리가 될 겁니다.
